src/python/dxpy/app_builder.py

```diff
@@ -212,58 +161,199 @@ def upload_resources(src_dir, project=None, folder='/', ensure_upload=False):
+        # otherwise, we will compress and upload the tarball.
+        
+        with tempfile.NamedTemporaryFile(suffix=".tar") as tar_fh:
+        
+            output_sha1 = hashlib.sha1()
+        
+            for dirname, subdirs, files in os.walk(resources_dir):
+                if not dirname.startswith(resources_dir):
+                    raise AssertionError('Expected %r to start with root directory %r' % (dirname, resources_dir))
+
+                # Add an entry for the directory itself
+                relative_dirname = dirname[len(resources_dir):]
+                dir_stat = os.lstat(dirname)
+                if not relative_dirname.startswith('/'):
+                    relative_dirname = '/' + relative_dirname
+
+                fields = [relative_dirname, str(dir_stat.st_mode), str(int(dir_stat.st_mtime * 1000))]
+                output_sha1.update(b''.join(s.encode('utf-8') + b'\0' for s in fields))
+                
+                # add an entry in the tar file for the current directory, but
+                # do not recurse!
+                subprocess.check_call(['tar', '-C', resources_dir, '-rf', tar_fh.name, '--no-recursion', '.' + relative_dirname])
+                
+                # Canonicalize the order of subdirectories; this is the order in
+                # which they will be visited by os.walk
+                subdirs.sort()
+                
+                # check the subdirectories for symlinks.  We should throw an error
+                # if there are any links that point outside of the directory (unless
+                # --force-symlinks is given).  If a link is pointing internal to 
+                # the directory (or --force-symlinks is given), we should add it 
+                # as a file.
+                for subdir_name in subdirs:
+                    dir_path = os.path.join(dirname, subdir_name)
+
+                    # If we do have a symlink,                     
+                    if os.path.islink(dir_path):
+                        # Let's get the pointed-to path to ensure that it is
+                        # still in the directory
+                        link_target = os.readlink(dir_path)
+                        is_local=(not os.path.isabs(link_target))
+                        
+                        dir_realpath=os.path.abspath(os.path.join(dirname, link_target))
+                        # make sure that the path NEVER extends outside the resources directory!
+                        d,l = os.path.split(link_target)
+                        link_parts = []
+                        while l:
+                            link_parts.append(l)
+                            d,l = os.path.split(d)
+                        curr_path = dirname
+                        for p in reversed(link_parts):
+                            curr_path = os.path.abspath(os.path.join(curr_path, p))
+                            is_local = (is_local and curr_path.startswith(resources_dir))
```

Comment:
To make this check work, `resources_dir` should also be made absolute. Otherwise you can end up with the following inappropriate error:

```
psung@exabase ~/source/cat $ dx build .
Error: Cannot include symlinks to directories outside of the resource directory.  u'./resources/a' points to directory u'/home/psung/source/cat/resources/b'
psung@exabase ~/source/cat $ ls -l resources
total 4.0K
lrwxrwxrwx 1 psung psung    1 Apr 25 21:58 a -> b
drwxrwxr-x 2 psung psung 4.0K Apr 25 22:04 b
```